indexer/src/main/java/org/uniprot/store/indexer/common/config/UniProtSolrClient.java

```diff
@@ -0,0 +1,160 @@
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.client.solrj.response.UpdateResponse;
+import org.apache.solr.common.params.ModifiableSolrParams;
+import org.uniprot.store.search.SolrCollection;
+import org.uniprot.store.search.document.Document;
+
+/**
+ * A wrapper of {@link SolrClient} which creates a {@link SolrClient} instance for each thread that
+ * it is used on. The purpose of this is to ensure that multi-threaded applications have multiple
+ * instances that can access Solr, e.g., improving writing throughput.
+ *
+ * <p>Created 10/07/19
+ *
+ * @author Edd
+ */
+@Slf4j
+public class UniProtSolrClient {
+    private final ThreadLocal<SolrClient> threadLocalSolrClients;
+    private final RepositoryConfigProperties config;
+
+    UniProtSolrClient(RepositoryConfigProperties config) {
+        this.config = config;
+        this.threadLocalSolrClients = ThreadLocal.withInitial(this::uniProtSolrClient);
+    }
+
+    public UniProtSolrClient(SolrClient solrOperations) {
+        this.config = null;
+        this.threadLocalSolrClients = ThreadLocal.withInitial(() -> solrOperations);
+    }
+
+    public <T extends Document> List<T> query(
+            SolrCollection collection, SolrQuery query, Class<T> documentClass) {
+        String collectionString = collection.name();
+        try {
+            QueryResponse response = threadLocalSolrClients.get().query(collectionString, query);
+            return response.getBeans(documentClass);
+        } catch (SolrServerException | IOException e) {
+            throw new SolrQueryRetrievalException(
+                    "Could not query from Solr collection [" + collectionString + "]", e);
+        }
+    }
+
+    public UpdateResponse saveBeans(SolrCollection collection, Collection<?> beans) {
+        String collectionString = collection.name();
+        try {
+            return threadLocalSolrClients.get().addBeans(collectionString, beans);
+        } catch (SolrServerException | IOException e) {
+            throw new SolrQueryRetrievalException(
+                    "Could not write documents to Solr collection [" + collectionString + "]", e);
+        }
+    }
+
+    public <T> Optional<T> queryForObject(
+            SolrCollection collection, SolrQuery solrQuery, Class<T> returnType) {
+        String collectionString = collection.name();
+        try {
+            QueryResponse response =
+                    threadLocalSolrClients.get().query(collectionString, solrQuery);
+            List<T> beans = response.getBeans(returnType);
+            if (beans.size() == 1) {
+                return Optional.of(beans.get(0));
+            } else {
+                return Optional.empty();
+            }
+        } catch (SolrServerException | IOException e) {
+            throw new SolrQueryRetrievalException(
+                    "Could not write documents to Solr collection [" + collectionString + "]", e);
```

Comment:
good point -- i followed the behaviour that we previously had, which was provided by spring data solr. I will keep the current behaviour because that is what we had previously, but i have changed the error message to make it clearer.